### PR TITLE
Cloudwatch logging verification & checks

### PR DESCRIPTION
**Pull request by hballuru**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description

Adding test case logic to detect whether cloudwatch logging is occurring as expected. This test case addresses a shortcoming with Sagemaker endpoints not generating any log events on Cloudwatch.

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [x] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py [hunk 1]
```diff
@@ -71,15 +78,15 @@ def test_mnist_eia(framework_version, ecr_image, instance_type, accelerator_type
...
+            'accelerator_type': accelerator_type,
 
         }
     invoke_pytorch_helper_function(ecr_image, sagemaker_regions, _test_mnist_distributed, function_args)
 
 
 
 def _test_mnist_distributed(
-        ecr_image, sagemaker_session, framework_version, instance_type, model_dir, mnist_script, accelerator_type=None
+        ecr_image, sagemaker_session, framework_version, instance_type, model_dir, mnist_script, accelerator_type=None, verify_logs: False
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:89]: `verify_logs=False`

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py [hunk 2]
```diff
@@ -118,3 +125,45 @@ def _test_mnist_distributed(
...
+    response = None
+    while response == None or response['status'] == 'Running':
+        print('Waiting for query to complete ...')
+        time.sleep(1)
+        response = client.get_query_results(
+            queryId=query_id
+        )        
+    recordsAvailable=bool(response['results'])
+    if not recordsAvailable:
+        print("Exception... No results")
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:160]: you can just raise exception here - will be easier to see logs

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py [hunk 3]
```diff
@@ -118,3 +125,38 @@ def _test_mnist_distributed(
...
+        logGroupName='/aws/sagemaker/Endpoints/'+endpoint_name,
+        logStreamName=logStreamName,
+        #startTime=int((datetime.today() - timedelta(days=2)).timestamp()),
+        #endTime=int(datetime.now().timestamp()),
+        limit=50,
+        startFromHead=True
+    )
+
+    print(new_response['events'])
+    recordsAvailable=bool(new_response['events'])
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:155]: this looks good for ensuring there are some logs. Now can we search for some specific messages that should be in the logs?

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py [hunk 4]
```diff
@@ -118,3 +124,45 @@ def _test_mnist_distributed(
...
+
+    log_events_response = client.get_log_events(
+        logGroupName='/aws/sagemaker/Endpoints/'+endpoint_name,
+        logStreamName=logStreamName,
+        limit=50,
+        startFromHead=True
+    )
+
+    #print(log_events_response['events'])
+    recordsAvailable=bool(log_events_response['events'])
```
Comment (kenny-ezirim) [inline test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:152]: # 152 Check log_events_response['events'] and raise exception if no logs were found. 
# 156 I don't think the print statement is necessary since you will be raising an exception. The exception message should say what happened.
# 153 & # 168 Maybe this print statement was necessary to debug the code but is it really needed?

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py [hunk 5]
```diff
@@ -118,3 +124,45 @@ def _test_mnist_distributed(
...
+def _check_for_cloudwatch_logs(endpoint_name):
+    client=boto3.client('logs')
+    identifyLogStream = client.describe_log_streams(
+        logGroupName='/aws/sagemaker/Endpoints/'+endpoint_name,    
+        orderBy='LastEventTime',
+        descending=True,
+        limit=5
+    )
+
+    logStreamName=identifyLogStream['logStreams'][0]['logStreamName']
```
Comment (kenny-ezirim) [inline test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:141]: Is an alternative but efficient way to retrieve logStreamName than indexing? This can lead to an IndexError if identifyLogStream['logStreams'] is None.

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py (hunk 4 above)
Comment (hballuru) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:152]: So for Line 153, it does seem the print is redundant, its merely a marker to see this statement appear in the logs. I realized the test case logging is not eloquent and you infact you don't see any of these print statements in the codebuild job log. Only when you go to the report and then click on each passed test case, is when you would see these. So if something has failed, it would be nice to see the line 153 printing so we know code until there is executed with no issues. So helpful when debugging. However I realize line 156 print statement is unnecessary. I will remove it.

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py (hunk 5 above)
Comment (hballuru) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py]: I thought of enclosing this in a try block. Bear in mind we do orderby last event time so if the describe_log_streams fails, then it can cause the indexing error like you mentioned, but that will fail the test and in this case I believe its ok? For now I am going to handle that exception and throw relevant information in the exception log

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py (hunk 2 above)
Comment (hballuru) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py]: done

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py (hunk 1 above)
Comment (hballuru) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py]: resolved

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py (hunk 3 above)
Comment (hballuru) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py]: Yes added logic for the same

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py [hunk 6]
```diff
@@ -118,3 +124,50 @@ def _test_mnist_distributed(
...
+    client=boto3.client('logs')
+    identifyLogStream = client.describe_log_streams(
+        logGroupName='/aws/sagemaker/Endpoints/'+endpoint_name,    
+        orderBy='LastEventTime',
+        descending=True,
+        limit=5
+    )
+
+    indexError=False;
+    try:
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:142]: why suppress the indexerror exception? You can remove the true/false logic and do something like this:

```
   try:
        logStreamName=identifyLogStream['logStreams'][0]['logStreamName']        
    except IndexError as e:
        raise RuntimeError(f"Unable to look up log streams in log stream {some_variable_that_identifies_log_stream_name}") from e
```

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py [hunk 7]
```diff
@@ -118,3 +124,50 @@ def _test_mnist_distributed(
...
 
         assert output.shape == (batch_size, 10)
+
+        #  Check for Cloudwatch logs
+        if verify_logs:
+            _check_for_cloudwatch_logs(endpoint_name)
+
+def _check_for_cloudwatch_logs(endpoint_name):
+    client=boto3.client('logs')
+    identifyLogStream = client.describe_log_streams(
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:134]: nit: Looks like we are mixing camel case and snake case most places. We generally use snake case in this codebase

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py [hunk 8]
```diff
@@ -118,3 +124,50 @@ def _test_mnist_distributed(
...
+    try:
+        logStreamName=identifyLogStream['logStreams'][0]['logStreamName']        
+    except IndexError:
+        indexError=True    
+
+    if(indexError):
+        raise Exception('Exception: Trying to lookup log streams from the given log group failed')
+
+    log_events_response = client.get_log_events(
+        logGroupName='/aws/sagemaker/Endpoints/'+endpoint_name,
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:151]: you probably can save log_group_name as a variable since it will likely be useful for error messages as well

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py [hunk 9]
```diff
@@ -118,3 +124,50 @@ def _test_mnist_distributed(
...
+        logStreamName=logStreamName,
+        limit=50,
+        startFromHead=True
+    )
+
+    #print(log_events_response['events'])
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:157]: remove debug statement

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py [hunk 10]
```diff
@@ -118,3 +124,50 @@ def _test_mnist_distributed(
...
+        print('INFO: Most recently logged events were found for the given log group & log stream... Now verifying that TorchServe endpoint is logging on cloudwatch')
+        check_for_torchserve_response = client.filter_log_events(
+            logGroupName='/aws/sagemaker/Endpoints/'+endpoint_name,
+            logStreamNames=[logStreamName], 
+            filterPattern='Torch worker started.',
+            limit=10,
+            interleaved=False
+        )
+        assert bool(check_for_torchserve_response['events'])
+        print('Success!')
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:173]: you can likely remove most of the print debug statements unless you think they would be useful. If you think useful, we should use python logging and set to "debug" level

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py [hunk 11]
```diff
@@ -118,3 +129,47 @@ def _test_mnist_distributed(
...
 
         assert output.shape == (batch_size, 10)
+
+        #  Check for Cloudwatch logs
+        if verify_logs:
+            _check_for_cloudwatch_logs(endpoint_name, sagemaker_session)
+
+def _check_for_cloudwatch_logs(endpoint_name, sagemaker_session):
+    client=sagemaker_session.boto_session.client('logs')
+    log_group_name='/aws/sagemaker/Endpoints/'+endpoint_name
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:139]: nit, generally we use f-strings.
```
"/aws/sagemaker/Endpoints/{endpoint_name}"
```
Additionally, it looks like there are a few places where formatting is off. Can you use this formatting tool - https://github.com/psf/black - on this file?

Once you install the package (via pypi), you can run the following command

```
black -l 120 test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py
```

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py [hunk 12]
```diff
@@ -118,3 +129,47 @@ def _test_mnist_distributed(
...
+
+
+    log_events_response = client.get_log_events(
+        logGroupName=log_group_name,
+        logStreamName=log_stream_name,
+        limit=50,
+        startFromHead=True
+    )
+    
+    records_available=bool(log_events_response['events'])
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:161]: Will `log_events_response['events']` be undefined or set to an empty data structure? if undefined, this will fail with key error. You can do something like log_events_response.get('events') which defaults to "None" if the key is not found to avoid this scenario

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py [hunk 13]
```diff
@@ -118,3 +124,45 @@ def _test_mnist_distributed(
         output = predictor.predict(data)
 
         assert output.shape == (batch_size, 10)
+
+        #  Check for Cloudwatch logs
+        if verify_logs:
+            _check_for_cloudwatch_logs(endpoint_name)
+
+def _check_for_cloudwatch_logs(endpoint_name):
+    client=boto3.client('logs')
```
[the remark this answers is not in the document]
Comment (hballuru) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py]: `client=sagemaker_session.boto_session.client('logs')`
Above should address regionalization. Let me know if that's not the case?

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py (hunk 6 above)
Comment (hballuru) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:148]: Thanks for the suggestion. This has been implemented.

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py (hunk 7 above)
Comment (hballuru) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py]: Changed all variables to snake case

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py (hunk 8 above)
Comment (hballuru) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py]: Done

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py (hunk 9 above)
Comment (hballuru) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py]: Done.. Removed all debug/print and replaced them with logger implementation

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py (hunk 10 above)
Comment (hballuru) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py]: Done.. Removed all debug/print and replaced them with logger implementation

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py (hunk 12 above)
Comment (hballuru) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:161]: No it will return false if there are no log events but should never return a null as long as there is an associated log stream that was retrieved a line above.

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py (hunk 11 above)
Comment (hballuru) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py]: Missed this somehow. I just rectified that and then reformatted.

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py [hunk 14]
```diff
@@ -118,3 +132,45 @@ def _test_mnist_distributed(
...
+        )
+        check_for_torchserve_response = client.filter_log_events(
+            logGroupName=log_group_name,
+            logStreamNames=[log_stream_name],
+            filterPattern="Torch worker started.",
+            limit=10,
+            interleaved=False,
+        )
+        assert bool(check_for_torchserve_response["events"])
+        LOGGER.info("Success!")
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py:176]: nit: make this a bit more specific - "cloudwatch log test successful" you can say

I would also make this a debug level statement, or remove

**Review by arjkesh — COMMENTED** [review]
Left one more nit comment. Outside of that, lgtm, you can revert toml in the next commit

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py (hunk 14 above)
Comment (hballuru) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_mnist.py]: Removed the logger statement

**Review by arjkesh — APPROVED** [review]
